### PR TITLE
EE-ES - gamelists - add sortname - fix

### PR DESCRIPTION
**Pull request by Langerz82**
@shantigilbert and @miwasp

Here is the final changes for testing. Sortmode should be working correctly, and also other modes. If an index is provided in the sortname ex: 
```
<sortname>001 - Blah</sortname>
```
Then indexes will be compared first, before alphabetic characters.

ref:
https://github.com/EmuELEC/emuelec-emulationstation/pull/34
https://github.com/EmuELEC/emuelec-emulationstation/pull/33

### CODE REVIEW

**Comment by Langerz82** [comment]
There is also 2 minor fixes which were unrelated to changes, but are so small they don't need a separate pull. 
1. The MetaID was specified as a digit rather than its enum name. 
https://github.com/EmuELEC/emuelec-emulationstation/blob/2ffacc53461f3a639e827efb096148d871a25022/es-app/src/MetaData.cpp#L300

2. Missing defines in some EmuELEC specific code.
https://github.com/EmuELEC/emuelec-emulationstation/blob/2ffacc53461f3a639e827efb096148d871a25022/es-app/src/guis/GuiMenu.cpp#L1618

**Comment by Langerz82** [comment]
I's a bit screwed atm, the sort list options dissappear, and sortName is slightly incorrect, give me some time to fix.

**Comment by miwasp** [comment]
I just tested this PR without the last 2 commits. Looks very promising ;-)

Normal systems (dreamcast, psx, ...) sort in the right order with name and sortname, both ascending or descending. If favourites are displayed on top of the list, entries with sortname are correctly ordered there. If you make an entry into a favourite it is immediately resorted. Really cool.

Two things i found so far:
1. sortorder "Release Date ascending" still displays the year as prefix while "Release Date descending" does not do this. I don't know if this has anything to do with these changes but with or without the prefix, it should be consistent.
2. The sortorder tag seems to be ignored in collections. If i make an entry to a favourite (and it is correctly sorted in the system view), then this entry is sorted with his name (instead of sortname) in the collection "Favourites".

**Comment by Langerz82** [comment]
@miwasp
> I just tested this PR without the last 2 commits. Looks very promising ;-)
> 
> Normal systems (dreamcast, psx, ...) sort in the right order with name and sortname, both ascending or descending. If favourites are displayed on top of the list, entries with sortname are correctly ordered there. If you make an entry into a favourite it is immediately resorted. Really cool.
> 
> Two things i found so far:
> 
> 1. sortorder "Release Date ascending" still displays the year as prefix while "Release Date descending" does not do this. I don't know if this has anything to do with these changes but with or without the prefix, it should be consistent.
> 2. The sortorder tag seems to be ignored in collections. If i make an entry to a favourite (and it is correctly sorted in the system view), then this entry is sorted with his name (instead of sortname) in the collection "Favourites".

1. This will need to go in a separate PR to avoid versioning issues as it's not quite related. It's a pretty simple task but I only include critical bugs that can potentially break code. It might be a good start project for you to start coding on. ;)
2. I think I understand what you mean. Maybe I'll fix that in separate PR, if I make more changes it may break something else again, which I do not want to make too many commits to this PR change.

**Comment by Langerz82** [comment]
Oh yeah and re-compile latest changes as it should fix indexing, and the sort list choosing.

**Comment by Langerz82** [comment]
Hey, I quickly made some changes to hopefully fix those 2 issues you mentioned, please test them again with latest changes, thanks.

**Comment by Langerz82** [comment]
Fixing Number 2 is allot more complicated as the collections sorting is more manual code. Feel free to make it a feature request and I might work on it when I get more free time.

**Comment by Langerz82** [comment]
@miwasp ok so I'm looking at the collections code. Did you want it to sort by system separately or by one massive sort? Currently I think it sorts seperately? Not too sure, can you check for me, if it sorts each system individually in the bigger lists?

**Comment by Langerz82** [comment]
@miwasp Collection sorting ready for testing. :]

**Comment by Langerz82** [comment]
I tested collection sorting briefly for the "all games" category and it seems to be working ok. Can you also test for custom collections too? I haven't checked yet.

**Comment by miwasp** [comment]
It‘s not how I want it to sort, i usually don‘t even use the sortorder tag. It‘s more how it worked until now and it should work the same way with the sortorder tag.
Every collection i have ever seen is one massive sort over all systems when sorted by name. The system is appended to the name in those lists. This way you can see e.g. on what systems a specific title is available. If i want it sorted per system i can look in the system directly or sort by system.

**Comment by shantigilbert** [comment]
I think this is going to be a hard one to please everyone, as you noted from the original code (Batocera) "SortName Metadata does not exist in batocera-emulationstation."   is not even used there (altho it is present in the [code](https://github.com/batocera-linux/batocera-emulationstation/blob/master/es-app/src/MetaData.h#L35)), how did it en up being used in EmuELEC-emulationstation?

I do not use that tag at all so I do not know how to help with the testing, I assume most gamelists.xml will not even have this, but sorting by name works now, and it didn't before.

**Comment by miwasp** [comment]
> how did it en up being used in EmuELEC-emulationstation?

You [merged](https://github.com/EmuELEC/emuelec-emulationstation/pull/23) it ;-)

**Comment by shantigilbert** [comment]
That explains it :p

**Comment by Langerz82** [comment]
Hey, I added in collections support for file sort filters, needs testing. I figured we might as well insert it in the fix too so sorting is more flexible.

**Comment by Langerz82** [comment]
Who knows maybe Batocera will even use these changes. I have made them pretty clean. Tbh I don't mind adding this feature, and it's not overly complex and taught me a few things. Before when I tried to separately sort sortname and name the code was messy as shit! lol. I think giving users the flexibility of being able to sort games however they want by using the sortname tag is a nice feature to have.

**Comment by Langerz82** [comment]
....and I'll be honest. Coding to me is like crack, I wana keep doing it to stay high! loool

**Comment by shantigilbert** [comment]
Yo are welcome to keep adding, modifying stuff for emuelec, I know we need more people like you on the team :)

**Comment by miwasp** [comment]
> 
> 
> I tested collection sorting briefly for the "all games" category and it seems to be working ok. Can you also test for custom collections too? I haven't checked yet.

Can you please explain what you mean with collection sorting? Are we still talking about the sortname tag?

What i meant when i said "the sortname tag seems to be ignored in collections" is the following:
You have for example a game with the name "aaa" in the system psx. You set a sortname "zzz" for this game.
If you are in the list of system psx, this game is displayed
- at the beginning with sortorder set to "name ascending"
- at the end with sortorder set to "sortname ascending"

If you are in the collection "All games", this game is displayed
- at the beginning with sortorder set to "name ascending"
- at the beginning with sortorder set to "sortname ascending"

This last one is wrong, it should be displayed at the end of the list (identical to the psx system view)
This was tested with all commits up to here.

**Comment by Langerz82** [comment]
hmm, sounds like the sorting may be getting ignored in collections, I'll look into it.

**Comment by Langerz82** [comment]
@miwasp hey what about custom collections? Is it doing the same thing regarding not sorting when sortname is added?
 I looked all over the collections file and couldn't find any code re-ordering the sort, I'll keep looking.

**Comment by miwasp** [comment]
> @miwasp hey what about custom collections? Is it doing the same thing regarding not sorting when sortname is added?

Yes, sortname is ignored there too.

**Comment by Langerz82** [comment]
Ok thanks for the update, I will look at it again tomorrow with fresh eyes.

File: es-app/src/FileSorts.cpp [hunk 1]
```diff
@@ -4,6 +4,8 @@
 #include "utils/StringUtil.h"
 #include "LocaleES.h"
 
+#include <climits>
```
Comment (shantigilbert) [inline es-app/src/FileSorts.cpp]: can you also wrap this in if _ENABLEEMUELEC

File: es-app/src/views/gamelist/GameNameFormatter.cpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ GameNameFormatter::GameNameFormatter(SystemData* system)
 
 	mShowYear =
 		mSortId == FileSorts::RELEASEDATE_ASCENDING ||
-		mSortId == FileSorts::RELEASEDATE_ASCENDING ||
+		mSortId == FileSorts::RELEASEDATE_DESCENDING ||
```
Comment (shantigilbert) [inline es-app/src/views/gamelist/GameNameFormatter.cpp:73]: And this as well, wrap it in if...else

**Comment by shantigilbert** [comment]
@Langerz82 please join Discord if you haven't or send me your Discord user :)

**Comment by Langerz82** [comment]
@miwasp It's definitely sorting by sortname, the logic in sortname just is not quite right. The list changes when sortname is applied, and thats a good indicator it's using different logic. The prob is you are right sortname is not taking proper priority which I am looking into.

File: es-app/src/views/gamelist/GameNameFormatter.cpp (hunk 2 above)
Comment (Langerz82) [reply es-app/src/views/gamelist/GameNameFormatter.cpp:73]: this part is a bug on batocera.

**Comment by Langerz82** [comment]
@shantigilbert Just gotta test this then ill switch OS, and give discord. ^^

**Comment by Langerz82** [comment]
ready for testing.

**Comment by miwasp** [comment]
The only difference i see so far is that the sortname instead of the name is now displayed in any system view.

Sorting is still wrong in collections. It's the same as i already described in my [comment](https://github.com/EmuELEC/emuelec-emulationstation/pull/35#issuecomment-932520413) 2 days ago.

**Comment by miwasp** [comment]
I think that we talk past another here. Let me try to explain it another way and please answer two questions then:
- Go to the collection "All games" and set sortorder to "Filename ascending".
- Take note of the first entry you see then. This entry is now our "testfile".
- Edit the gamelist.xml of the system which contains the "testfile" and set a sortname "zzz-whatever" for it.
- Save that gamelist.xml and restart ES
- This is now our testcase

- Go to the system view which contains your "testfile" and set sortorder to "Filename ascending". This must be the first entry then. This works.
- Stay in this system view and set sortorder to "Sortname ascending". Your "testfile" must be the last entry then. This works too.
- Now go to the collection "All games" and set sortorder to "Filename ascending". Your "testfile" must be the first entry. This works too.
- Stay in collection "All games" and set sortorder to "Sortname ascending". Your "testfile" must be the last entry then but it is not. It is still the first entry.

Now my two questions:
1. Do you understand it the same way, should it work this way in your opinion?
2. Can you reproduce this testcase?

**Comment by Langerz82** [comment]
Yeah I know sortname is not being compared for some reason. Odd let me check.

**Comment by Langerz82** [comment]
@shantigilbert , @miwasp 
Finally got it working, the issue was the string was going out of scope making it point to an invalid memory address. Let me clean up the code, but yeah from the "init sortName" commit it should work.

**Comment by Langerz82** [comment]
@miwasp  Testing is ready, I made a couple of fixes should compile smooth now and have sortName in collection lists if it's provided in the gamelist.xml file.

**Comment by miwasp** [comment]
Congrats!
Tested it with system view, predefined collections and custom collections, with and without favourites, filename and sortname sorting, ascending and descending, ...
Everything works like it should and its consistent over all views.
Good job.

**Comment by Langerz82** [comment]
@shantigilbert Give it a couple days I want to thoroughly test it, make sure everything is working good before merging.

**Comment by Langerz82** [comment]
I have noticed one bug. When using Jump to letter it defaults to filename, ascending.

**Comment by Langerz82** [comment]
jump to letter is not working for sortname, I gotta fix some of the code up maybe later today.

**Comment by Langerz82** [comment]
Jump to letter should be working now for sortname too. I've added sorting for prefix indexes for just names too, as I think it should be the same, example "001 - Blah" displaying before "1 - Blah".

Final test please before merge. :dancers:

**Comment by shantigilbert** [comment]
Is this good to merge?

**Comment by Langerz82** [comment]
@miwasp Can you do us a huge favour and test this latest change? I've tested ordering for both name and sortname, and jump to letter but could do with an extra tester just to be sure.

**Comment by miwasp** [comment]
Sure, i'm compiling right now...

**Comment by miwasp** [comment]
So i made several tests now and i found just one minor issue:
If you sort by "Filename, ascending" and open the view options, the setting "Jump to letter" always shows the letter where you were in the list when you opened the options.
If you sort by "Sortname, ascending" and open the view options, the setting "Jump to letter" always shows the letter "O", no matter where in the list you were when you opened the options.

**Comment by miwasp** [comment]
Oh, and in MetaData.cpp in line 310 there is a LOG-message. I think this was for debugging purposes and is no longer needed?

**Comment by Langerz82** [comment]
I'll take a look and make a revision prob tomorrow when I got some free time. Thanks @miwasp .

**Comment by Langerz82** [comment]
I've found a bug with new and old code too. When sorting by filename or sortname descending it does not jump to letter properly. I'll have to look into this further before merging.

**Comment by Langerz82** [comment]
@miwasp can you do a final test before we merge this ? Jump too letter should work for both filename/sortname ASC and DESC.

**Comment by miwasp** [comment]
Looks good to me. I think this is ready to merge.

**Comment by Langerz82** [comment]
Ok cool. I've tested this enough to say merge too.

@shantigilbert whenever you are ready. ^^

**Comment by shantigilbert** [comment]
@Langerz82 @miwasp Thank you both! I appreciate the work and PRs!

**Pull request merged**